hspec-core/src/Test/Hspec/Core/Formatters/Internal.hs

```diff
@@ -259,21 +278,12 @@ htmlSpan :: String -> FormatM a -> FormatM a
 htmlSpan cls action = write ("<span class=\"" ++ cls ++ "\">") *> action <* write "</span>"
 
 withColor_ :: SGR -> FormatM a -> FormatM a
-withColor_ color (FormatM action) = do
-  useColor <- getConfig formatConfigUseColor
-  h <- getHandle
-
-  FormatM . StateT $ \st -> do
-    bracket_
-
-      -- set color
-      (when useColor $ hSetSGR h [color])
-
-      -- reset colors
-      (when useColor $ hSetSGR h [Reset])
-
-      -- run action
-      (runStateT action st)
+withColor_ color action = do
+  oldColor <- gets stateColor
+  modify (\state -> state { stateColor = Just color })
+  x <- action
+  modify (\state -> state { stateColor = oldColor })
+  return x
```

Comment:
I really like this.  It's great how we get arbitrary color nesting for free with this approach :clap: